ext/drmgr/drmgr.c

```diff
@@ -2497,3 +2510,134 @@ drmgr_disable_auto_predication(void *drcontext, instrlist_t *ilist)
     instrlist_set_auto_predicate(ilist, DR_PRED_NONE);
     return true;
 }
+
+/***************************************************************************
+ * EMULATION
+ */
+
+/*
+ * Constants used when accessing emulated instruction data with the
+ * drmgr_get_emulatd_instr_data() function. Each constant refers to an element
+ * of the emulated_instr_t struct.
+ */
+typedef enum {
+    DRMGR_EMUL_INSTR_PC, /* The PC address of the emulated instruction. */
+    DRMGR_EMUL_INSTR,    /* The emulated instruction. */
+} emulated_instr_data_t;
+
+/* Reserve space for emulation note values */
+static void
+drmgr_emulation_init(void)
+{
+    note_base_emul = drmgr_reserve_note_range(DRMGR_NOTE_EMUL_COUNT);
+    ASSERT(note_base_emul != DRMGR_NOTE_NONE, "failed to reserve emulation note space");
+}
+
+/* Get note values based on emulation specific enums. */
+static ptr_int_t
+get_emul_note_val(int enote_val)
+{
+    return (ptr_int_t)(note_base_emul + enote_val);
+}
+
+/* Write emulation data to a label's data area. */
+static void
+set_emul_label_data(instr_t *label, int type, ptr_uint_t data)
+{
+    dr_instr_label_data_t *label_data = instr_get_label_data_area(label);
+    ASSERT(label_data != NULL, "failed to find label's data area");
+
+    ASSERT(type >= DRMGR_EMUL_INSTR_PC && type <= DRMGR_EMUL_INSTR,
+           "type is invalid, should be an emulated_instr_data_t");
+    label_data->data[type] = data;
+}
+
+/* Read emulation data from a label's data area. */
+static ptr_uint_t
+get_emul_label_data(instr_t *label, int type)
+{
+    dr_instr_label_data_t *label_data = instr_get_label_data_area(label);
+    ASSERT(label_data != NULL, "failed to find label's data area");
+
+    ASSERT(type >= DRMGR_EMUL_INSTR_PC && type <= DRMGR_EMUL_INSTR,
+           "type is invalid, should be an emulated_instr_data_t");
+    return label_data->data[type];
+}
+
+/* A callback function to free the emulated instruction represented by the label.
+ * This will be called when the label is freed.
+ */
+static void
+free_einstr(void *drcontext, instr_t *label)
+{
+    instr_t *einstr = (instr_t *)get_emul_label_data(label, DRMGR_EMUL_INSTR);
+    instr_destroy(drcontext, einstr);
+}
+
+/* Set the start emulation label and emulated instruction data */
+DR_EXPORT
+bool
+drmgr_insert_emulation_start(void *drcontext, instrlist_t *ilist, instr_t *where,
+                             emulated_instr_t *einstr)
+{
+    ASSERT(sizeof(emulated_instr_t) <= sizeof(dr_instr_label_data_t),
+           "label data area is not large enough to store emulation data");
```

Comment:
This can go in drmgr_emulation_init(), right?  2 benefits to moving it there: 1) It will fire even if this code is not exercised, so we'll catch errors more quickly on changes to these structs; 2) the check is static: no reason to re-check on every call here.

Reply:
Good point. Fixed.